### PR TITLE
wip: create descriptors for public schema

### PR DESCRIPTION
**Pull request by RichardJCai**

### CODE REVIEW

File: pkg/sql/catalog/bootstrap/metadata.go [hunk 1]
```diff
@@ -115,21 +115,19 @@ func (ms MetadataSchema) GetInitialValues() ([]roachpb.KeyValue, []roachpb.RKey)
...
 				ret,
 				roachpb.KeyValue{
 					Key:   catalogkeys.MakeDatabaseNameKey(ms.codec, desc.GetName()),
 					Value: value,
 				},
-				roachpb.KeyValue{
-					Key:   catalogkeys.MakePublicSchemaNameKey(ms.codec, desc.GetID()),
-					Value: publicSchemaValue,
-				})
+			)
```
Comment (postamar) [inline pkg/sql/catalog/bootstrap/metadata.go:130]: This whole if/else block can be replaced with
```
ret = append(ret, roachpb.KeyValue{Key: catalogkeys.EncodeNameKey(ms.codec, desc), Value: value})
```
as we now have a 1-to-1 mapping from descriptor to namespace entry.

**Review by postamar — COMMENTED** [review]
Not sure if this is helpful but here's what I thought a migration would look like:
https://gist.github.com/postamar/bd11f1abe032eb6887d78643c94ad927
https://gist.github.com/postamar/b9da9ef2f3714b4dd0f1510900074365

I don't have anything else that's meaningful to share, I'm afraid. The migration does the change using the catalog API but I've been wondering if it would be better to do it via SQL statements, as in: for each database create a new `crdb_internal_public_migration` schema or something and reparent everything from `public` to that new schema using `ALTER TABLE ... SET SCHEMA` and then at the very end rename that schema to `public` somehow.
I'm sorry this is vague but I didn't go further because I was stumped by the bootstrapping as well. I got a `WriteTooOldError` but I hadn't added zone configs for the new system schema descriptor so perhaps I did it wrong. For what it's worth I also thought that using ID `29` for the system schema was a good idea.

**Comment by RichardJCai** [comment]
I think I've managed to figure out the bootstrapping issue.

Now another question I have is if a startup migration is better in this case, version gating all the schema resolution logic seems tedious.

**Comment by ajwerner** [comment]
> Now another question I have is if a startup migration is better in this case, version gating all the schema resolution logic seems tedious.

In principle I'm opposed to the startup migration. We have so many things which scan the namespace table and then look up descriptors and that code is not expecting a desc. Counting on them doing the right thing seems hard. 

Also it's an O(descriptors) migration. So, yeah, generally I don't love that idea. The flip side is, as you pointed out, figuring out all the spots we treat it differently or assume things and combining version gating with other techniques.

**Pull request closed without merging**